### PR TITLE
feat: configure and provision local embedding models

### PR DESCRIPTION
**Pull request by BobdenOs**
## Summary

- replace the handwritten WordPiece implementation with `@huggingface/tokenizers@0.1.3`
- expose `@huggingface/tokenizers` as an optional peer dependency and fail with an installation hint when `ai-sqlite` uses it without the package installed
- require an explicit embedding model; there is no built-in/default model
- keep runtime configuration limited to `embedding: { model, directory? }`
- discover compatible Hugging Face ONNX artifacts and embedding metadata automatically from the model name
- provision missing project-local models on demand into `<cds.root>/.cds/models/<model>` after logging a warning
- support explicit local/shared provisioning with `npx @cap-js/ai install-model <model> [--directory <root>]`
- generate a pinned `embedding.lock.json` containing the resolved revision, files, checksums, dimensions, token limit, pooling, and normalization
- validate staged models through ONNX Runtime before publishing them and cleanly dispose validation/runtime sessions
- embed one model input window and truncate longer input; applications should split long documents and store one vector per chunk

## Runtime configuration

```json
{
  "cds": {
    "requires": {
      "db": {
        "kind": "ai-sqlite",
        "embedding": {
          "model": "foo/bar"
        }
      }
    }
  }
}
```

Without `directory`, the effective installation is:

```text
<cds.root>/.cds/models/foo/bar
```

For a shared pre-installed cache:

```sh
npx @cap-js/ai install-model foo/bar --directory ~/.cds/models
```

```json
{
  "embedding": {
    "model": "foo/bar",
    "directory": "~/.cds/models"
  }
}
```

Configured directories are verified but remain offline/read-only during application startup.

## Model compatibility

- resolves Hugging Face revisions to immutable commits
- supports conventional `onnx/model.onnx`, including external ONNX data files
- derives dimensions and bounded token length from model/tokenizer metadata
- derives pooling and normalization from an unambiguous Sentence Transformers module chain
- validates input/output names, tensor shape/type, and dimensions through a startup probe
- rejects missing or ambiguous model metadata instead of guessing defaults

## Validation

- embedding, discovery, provisioning, `ai-sqlite`, and knowledge-graph tests: 68/68 passing
- `npx -y eslint@10 .`
- `npx -y prettier@3 --check .`
- `git diff --check`

## Stack

- includes merged #55
- rebased onto `AISQLiteService`, including #49's triple-store integration and #58's single-window behavior
- `onnxruntime-node` remains pinned to `1.20.1`; broader compatibility is tracked in #54

### Category: Feature / Refactor

- [x] Added relevant changelog documentation

### CODE REVIEW

**Comment by sjvans** [comment]
Summary of the update:

- Rebased the tokenizer work onto `AISQLiteService` and kept `Xenova/all-MiniLM-L6-v2` as the zero-configuration default.
- Replaced the handwritten WordPiece tokenizer with `@huggingface/tokenizers@0.1.3`, using the model-provided tokenizer artifacts instead of duplicating tokenizer behavior.
- Reworked long-text chunking so every chunk retains tokenizer-derived special tokens, attention masks, and token type IDs without depending on tokenizer-side truncation.
- Replaced the shared embedding state with a runtime scoped to each `ai-sqlite` service, allowing different services to use different compatible encoder configurations safely.
- Added explicit custom-model descriptors covering repository, immutable revision, dimensions, maximum sequence length, artifact roles, expected sizes, and SHA-256 checksums. MiniLM remains the default when no descriptor is supplied.
- Kept downloads bounded, checksum-verified, atomic, and cache-isolated by repository, revision, and artifact-set digest.
- Added startup validation for ONNX input names, output names, tensor types, tensor shapes, and configured dimensions. Compatible models may use `input_ids`, `attention_mask`, and `token_type_ids`, and support `mean`, `cls`, or already-pooled output handling.
- Preserved the HANA-compatible `VECTOR_EMBEDDING` signature. `model_and_version` remains informational and does not dynamically select or download a model during a query.
- Expanded the unit and integration coverage for tokenizer boundaries, long-input chunking, custom descriptors, cache isolation, pooling, normalization, optional inputs, invalid model contracts, and per-service runtime isolation.
- Updated the README and changelog with the supported custom-model contract and security requirements.

Verification on commit `394af11`: all 38 tests pass, including 24 focused tokenizer/vector tests; ESLint, Prettier, and `git diff --check` pass.

`onnxruntime-node` remains pinned to `1.20.1`; investigation of later versions and the private synchronous runtime binding is tracked separately in #54.

File: lib/vector_embedding/embedding.js [hunk 1]
```diff
@@ -3,208 +3,341 @@ import path from 'path';
...
-const MODEL = {
+const STANDARD_INPUT_NAMES = new Set(['input_ids', 'attention_mask', 'token_type_ids']);
+
+const DEFAULT_MODEL = Object.freeze({
   repository: 'Xenova/all-MiniLM-L6-v2',
   revision: '751bff37182d3f1213fa05d7196b954e230abad9',
-  files: [
-    {
+  dimensions: 384,
+  maxLength: 128,
```
Comment (PDT42) [inline lib/vector_embedding/embedding.js]: This will make the tokenizer yield chunks of up to 128 tokens; This could be parts of sentences.

File: lib/sqlite/AISQLiteService.js [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
 import SQLiteService from '@cap-js/sqlite';
-import { initializeEmbedding, vector_embedding } from '../vector_embedding/index.js';
+import { createEmbeddingRuntime } from '../vector_embedding/embedding.js';
+
+const LOG = cds.log('@cap-js/ai');
 
 export default class AISQLiteService extends SQLiteService {
   async init() {
-    await initializeEmbedding();
+    this._embeddingRuntime = await createEmbeddingRuntime(this.options.embedding);
```
Comment (PDT42) [inline lib/sqlite/AISQLiteService.js]: AFAIK this means the user has to provide a model config (same as the `DEFAULT_MODEL` all of the keys) in their `.cdsrc`. IMO we should make some of the options and validations optional: E.g. if the user provides a sha for the model they want to use, we make sure it matches - but if the user does not provide a value, we shouldn't fail ... 

IMO `validateModelDescriptor` is overzealous.

File: lib/vector_embedding/embedding.js [hunk 3]
```diff
@@ -3,208 +3,341 @@ import path from 'path';
...
+  });
 
-    embeddings.push(pooledEmbedding);
+  const combined = new Float32Array(model.dimensions);
+  for (const embedding of embeddings) {
+    for (let index = 0; index < combined.length; index++) combined[index] += embedding[index];
+  }
+  if (embeddings.length > 1) {
+    for (let index = 0; index < combined.length; index++) combined[index] /= embeddings.length;
   }
```
Comment (PDT42) [inline lib/vector_embedding/embedding.js]: This will still use averages to merge embeddings from the same text 😞 There is no real way around this: 
If the input's length exceeds the max token window of the model, the model would [truncate the input](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2#:~:text=By%20default%2C%20input%20text%20longer%20than%20256%20word%20pieces%20is%20truncated.). 

We either choose to accept information loss from truncation by the model, or try to chunk the input and deal with the fact that one input will yield multiple result vectors. We could store multiple vectors pointing to the same ID, but enabling something like that generically seems like a stretch; Alternatively, we need to either, aggregate or drop embeddings. 

To get rid of averages, we can consider: 
- Dropping embeddings (keeping the full text)
- Use smarter aggregation e.g. "Weighted mean by token count" ... see if that does better

IMO, we should just skip pooling over chunks; drop embeddings or let the model truncate.

File: lib/vector_embedding/model-utils.js [hunk 4]
```diff
@@ -106,136 +231,28 @@ async function downloadModelIfNeeded(modelDir, model, options) {
...
 
-  const session = await InferenceSession.create(modelPath);
   return {
-    session,
-    tokenizer: {
-      vocab,
-      maxLength,
-      normalizer: tokenizerJson.normalizer ?? {}
-    }
+    session: await InferenceSession.create(modelPath),
```
Comment (PDT42) [inline lib/vector_embedding/model-utils.js:727]: This explains the existence of `InferenceSession` as a compatibility wrapper. If we'll ever get a usable session straight from `onnx`, we can put it here 🙈

File: lib/vector_embedding/embedding.js (hunk 3 above)
Comment (PDT42) [reply lib/vector_embedding/embedding.js]: For reference: I asked Claude what 300 tokens (i.e. slightly more than the 256 hardcap of xenova) of input might look like, when assuming word-piece-tokenization; When read from CAPire. Claude came up with this example:

> ### Auto-Exposed and Generated Entities
> 
> In general, a service actually exposes more than the explicitly modeled entities from the CDS service model. This stems from the fact that the compiler auto-exposes entities for the sake of completeness, for example, by adding composition entities. Another reason is generated entities for localization or draft support that need to appear in the service. Typically, such entities don't have restrictions. The emerging question is, how can requests to these entities be authorized?
> 
> Basically, users with the Supporter role aren't restricted, whereas authenticated users can only read the Components. But what about the auto-exposed entities such as IssuesService.Issues and IssuesService.Categories? They could be a target of an (indirect) request, but none of them are annotated with a concrete restriction. In general, the same also holds for service entities that are generated by the compiler, for example, for localization or draft support.
> 
> ---
> 
> To close the gap with auto-exposed and generated entities, the authorization of such entities is delegated to a so-called authorization entity, which is the last entity in the request path that bears authorization information — that is, an entity that fulfills at least one of the following properties:
> 
> Explicitly exposed in the service
> Annotated with a concrete restriction
> Annotated with @cds.autoexpose

I drew a line, approximately where I would estimate the 256 token limit to be located. It's actually more than I'd expected, but I still don't think that we should use the average of the two, to describe the whole ...

**Review by PDT42 — DISMISSED** [review]
IMO this should replace: https://github.com/cap-js/ai/pull/53
 
- The model configuration should be updated to require only, `model`, local dir where to store the model, `normalization`, `pooling`; could consider keeping the others for consistency checks
- I'd like more readable test titles, please 😅; Currently the titles don't tell me what the test expects to happen and doesn't allow me to agree or disagree with that expectation

**Comment by PDT42** [comment]
After some confusion about the `@cap-js/sqlite` version and figuring out, I have to supply `cosine_similarity` ...

<details>
<summary>See Details</summary>
Edited the ai node module with: 

```js
dbc.function('COSINE_SIMILARITY', deterministic, (a, b) => cosineSimilarity(toFloatArray(a), toFloatArray(b)))

...

function toFloatArray(vector) {
  if (vector == null) return null
  if (vector instanceof Float32Array) return Array.from(vector)
  if (Buffer.isBuffer(vector)) return JSON.parse(vector.toString('utf8'))
  if (vector instanceof Uint8Array) return JSON.parse(Buffer.from(vector).toString('utf8'))
  if (typeof vector === 'string') return JSON.parse(vector)
  if (Array.isArray(vector)) return vector
  throw new Error(`Unsupported vector type: ${typeof vector}`)
}

function cosineSimilarity(a, b) {
  if (a == null || b == null) return null
  let dot = 0, normA = 0, normB = 0
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i]
    normA += a[i] * a[i]
    normB += b[i] * b[i]
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB)
  return denom === 0 ? 0 : dot / denom
}
```
</details>

I was able to make this work in https://github.tools.sap/cap/know-cap/pull/20 👌

File: package.json [hunk 5]
```diff
@@ -20,6 +20,9 @@
     "lib",
     "srv"
   ],
+  "dependencies": {
+    "@huggingface/tokenizers": "0.1.3"
```
Comment (sjvans) [inline package.json]: make optional peer dep

**Review by PDT42 — APPROVED** [review]
